virtcontainers/sandbox.go

```diff
@@ -1873,14 +1873,28 @@ func (s *Sandbox) updateResources() error {
 	// If the CPUs were increased, ask agent to online them
 	if oldCPUs < newCPUs {
 		vcpusAdded := newCPUs - oldCPUs
-		if err := s.agent.onlineCPUMem(vcpusAdded, true); err != nil {
+		if err := s.agent.onlineCPUMem(vcpusAdded, true, false); err != nil {
 			return err
 		}
 	}
 	s.Logger().Debugf("Sandbox CPUs: %d", newCPUs)
 
 	// Update Memory
 	s.Logger().WithField("memory-sandbox-size-byte", sandboxMemoryByte).Debugf("Request to hypervisor to update memory")
+	reqMemMB := uint32(sandboxMemoryByte >> utils.MibToBytesShift)
+	currentMemMB := s.hypervisor.getMemorySize()
+
+	// If request hotplug memory size larger than utils.MaxHotplugMemMBOnceTime,
+	// inorder to avoid hotplug memory oom problem, we need to hotplug large memory
+	// with two steps. First, hotplug utils.MaxHotplugMemMBOnceTime size memory into
+	// guest and wait all hotplug memory online. Then, hotplug the left unplugged memory
+	// into the guest
+	if currentMemMB < reqMemMB && (reqMemMB-currentMemMB) > utils.MaxHotplugMemMBOnceTime {
+		if err := s.beforeHotplugHugeMem(currentMemMB); err != nil {
```

Comment:
I guess this makes sense as you've already queried the hypervisor memory so why not pass it as an argument. But that function could do that itself as it's a method.